### PR TITLE
Update Decaffeination Guidelines

### PR DESCRIPTION
**Pull request by confused-Techie**
Due to the recent discovery of CoffeeScript 2.x tooling, breaking our CoffeeScript made for 1.x tooling, I have updated the guidelines and pointed out tools that can be used to successfully decaffeinate code. 

Please feel free to point out any bad grammar or things that are unclear.

### CODE REVIEW

**Comment by Daeraxa** [comment]
Below is how I would have written that section. Feel free to agree or disagree:

>As for how to successfully decaffeinate the core of the editor (or any other repos/packages in use), a great template can be found in [`pulsar-edit/pulsar` PR #13](https://github.com/pulsar-edit/pulsar/pull/13). 
What you must keep in mind is that when Atom was originally created they used tooling for CoffeeScript 1.x whereas much of the tooling that's available now is CoffeeScript 2.x. This means is that in order to decaffeinate without breaking changes, the tooling for version 1.x **MUST** be used otherwise there may be unknown ramifications.
We have found that the best tool to use is [`decaffeinate/decaffeinate` Version 4.0.0](https://github.com/decaffeinate/decaffeinate/releases/tag/v4.0.0). This is the stable release of their decaffeination tooling that uses CoffeeScript 1.x. Anything beyond that could cause breaking changes but as long as you are able to ensure that the tooling you use focuses on CoffeeScript 1.x then there should be few issues.
>
>After using this tool, while the JavaScript generated may not be the prettiest, it will at least be functional. From that functional JavaScript you can change any declarations of `var` to `let` or `const` as needed. This can always be taken further if needed but this alone should be enough to successfully decaffeinate. 
>
>If you'd like to see some discussion about this change in how we decaffeinate, the initial problem was discovered during a [Discussion on Discord](https://discord.com/channels/992103415163396136/992103415163396139/1018628639404863550) and the solution was found on a [PR here](https://github.com/pulsar-edit/pulsar/pull/45) by a long-time member of both Atom and Pulsar.

**Comment by confused-Techie** [comment]
Great suggestions @Daeraxa, I've implemented them, adding a comma here or there to make it feel right, but let me know what you think.

**Comment by Daeraxa** [comment]
Looks fine to me

**Comment by DeeDeeG** [comment]
I don't have anything formal or any exact wording to offer.

But from my experience with apm, I would say the following:

(In short):
- It might be worth checking each project to see if it uses CoffeeScript 1.x or 2.x tooling, usually by seeing which version of the `coffee-script` or `coffeescript` npm package it specifies in `package.json`.
- There is an official `coffee` CLI program by the CoffeeScript authors, which people may want to consider as an alternative to the "decaffeinate" tool.

<details><summary>Details (click to expand)</summary>

- You can usually tell whether a project uses CoffeeScript 1.x or 2.x based on what it requires in its package.json (usually under devDependencies).
  - Example: https://github.com/atom/apm/blob/8538e0f82522be3793c947f7c63408124a4e8613/package.json#L58
  - Probably only 1.x in the Atom repos, but I'm not fully 100% sure about that. Worth checking each project just to be safe?
- Rather than the decaffeinate tool, which I am just now hearing about, I think you can use the official `coffee` CLI program (https://www.npmjs.com/package/coffeescript/v/1.12.7) and (recursively I think?) "decaffeinate" a directory of .coffee files to .js JavaScript files.
  - I would personally be inclined to trust that `coffee`, although I don' really know the `decaffeinate` tool or have any ability to compare them really at the moment.
  - Examples of using the `coffee` command can be found in the scripts of apm's `package.json`: https://github.com/atom/apm/blob/8538e0f82522be3793c947f7c63408124a4e8613/package.json#L23

</details>

**Comment by confused-Techie** [comment]
@DeeDeeG thanks for the insights. Added this in, may have required a bit of a rewrite, but still tried to use as many of the suggestions implemented previously. Thanks!

**Comment by DeeDeeG** [comment]
Here's a pass of some copy editing, wording changes for clarity/readability, some changes for brevity/voice/style (i.e. totally subjective but hopefully people like these style changes), markdown formatting with numbered bullets and subheadings to break up the length a bit and emphasize key information, a couple clarifications on technical details to the best I understand the tech.

Hope this is good. This is a suggestion so feel free to improve it or not take all these changes.

<details><summary>More thoughts (click to expand):</summary>

I guess this content could be shorter so it could be read quicker. I genuinely don't know of any specific CoffeeScript 2.x in the repos, so that section maybe can just outright say "no 2.x only 1.x", but I can't be confident in that, really, so it's good as-is IMO? On the other hand, regarding length vs brevity, this gives tons of context, which could be nice for total newcomers. I hope that, ideally, folks can skim for what they're looking for and ignore the rest if it's "too much".

</details>

(Note for grabbing these changes: Admins of this repo should be able to "three bubbles menu" --> "edit" my comment, and from there you can copy-paste the markdown out of my comment directly if these changes look good.)

## Decaffeination

A goal for the core of the editor is to 'decaffeinate' (to remove CoffeeScript and replace it with JavaScript). This goal comes largely from a group decision, considering that very few of our core team are familiar with or prefer CoffeeScript over JavaScript. Additionally a big push of `Pulsar` is to lower the barrier to entry for new contributors to be as low as possible. And using fewer languages assists in that goal.

As for how to successfully Decaffeinate the core of the editor (or any other repos/packages in use), a great template for successfully doing so can be seen on [`pulsar-edit/pulsar` PR #13](https://github.com/pulsar-edit/pulsar/pull/13).

### CoffeeScript 1.x or 2.x

What you must keep in mind is that when Atom was originally created, its authors mostly used tooling for CoffeeScript 1.x, whereas much of the latest tooling that's available is CoffeeScript 2.x based. This means that in order to decaffeinate successfully the tooling made for version CoffeeScript 1.x **MUST** be used, otherwise there may be JavaScript code produced that behaves significantly differently than the original CoffeeScript. (This could cause bugs and errors in the resulting code.)

Nevertheless, there may be some CoffeeScript 2.x code somewhere among Atom's codebase. In order to accurately determine which is being used within the repo you are working on, the best way to do so is to look at that repo's `package.json` `devDependencies` or sometimes it's `dependencies`.

Within those fields look for the `coffee-script` or `coffeescript` package to see if its versions are 1.x or 2.x.

In general it'll probably be 1.x tooling used.

### How to Decaffeinate

Now once the version is known, you have some options in the best tools to use:

1) Some of the members prefer using the official [`coffee` program](https://www.npmjs.com/package/coffeescript), with [1.12.7](https://www.npmjs.com/package/coffeescript/v/1.12.7) being recommended for CoffeeScript 1.x tooling and latest being fine for 2.x tooling.

(If using the official Coffee program, here is an example from one of the [scripts within `atom/apm`](https://github.com/atom/apm/blob/8538e0f82522be3793c947f7c63408124a4e8613/package.json#L23). You can run `coffee --compile --output lib src` to decaffeinate all .coffee files in the `src/` folder, and place them into `lib/` as .js files.)

2) Otherwise other members prefer [`decaffeinate/decaffeinate` Version 4.0.0](https://github.com/decaffeinate/decaffeinate/releases/tag/v4.0.0) for CoffeeScript 1.x tooling, and [`decaffeinate/decaffeinate` latest](https://github.com/decaffeinate/decaffeinate) for 2.x tooling.

After using either tool, while the JavaScript generated may not be the prettiest, it will at least be functional. From that functional JavaScript you can change any declarations of `var` to `let` or `const` as needed. Code improvement can always be taken further if needed, but this alone should be enough to successfully decaffeinate.

If you'd like to see some discussion about this change in how we decaffeinate, the initial problem was discovered during a [Discussion on Discord](https://discord.com/channels/992103415163396136/992103415163396139/1018628639404863550) and the solution was found on a [PR here](https://github.com/pulsar-edit/pulsar/pull/45) by a longtime member of both Atom and Pulsar.

### Testing

After something has been decaffeinated, it's asked that you test as much as possible. And if testing is not always possible, all new PR's on `pulsar-edit/pulsar` receive GitHub Actions testing automatically.

**Comment by fabianfiorotto** [comment]
It doesn't matter what tool you use. You always need to review the code line by line if you want to simplify it. Any compiler you use will give you a code that works but it's hard to read.  So we have two options. 1. leave the code as is and get a code that will work but nobody will want to maintain or 2. adapt the code manually and get a code that can have errors but is easy to read and maintain.

**Comment by softcode589** [comment]
@confused-Techie Should I add a review for everyone else who hasn't took a turn, just so that a lot of eyes are looking at this?

**Comment by confused-Techie** [comment]
> @confused-Techie Should I add a review for everyone else who hasn't took a turn, just so that a lot of eyes are looking at this?

Totally happy with that, since it is a pretty big guidelines change, and likely one that'll effect a lot of work people do. Although many changes have yet to be implemented

**Comment by confused-Techie** [comment]
Copied from another comment from @fabianfiorotto which I think applies exceptionally well as a suggestion here:

Most of the lines look as if they are taken directly from the compiler. We need to change them to remove the unneeded things in order to make the code more readable and easy to mantain. That includes

- Replace `var` with `const` and `let` if it's possible
- Replaced nested ternaries operators with `if` statements
- Remove `return` when it's not needed
- Remove result arrays when they aren't needed.
- Replace `for` loops with `for-of` loops
- Use arrow functions to write functions in a single line when it's possible
- Remove unneeded `ref` variables
- Remove [do](https://devdocs.io/coffeescript~1/index#loops) blocks and use let or const instead

File: project-birth/CONTRIBUTING-DURING-START.md [hunk 1]
```diff
@@ -76,7 +76,23 @@ When rebranding source code that creates or displays strings the end user will s
...
+As for how to successfully Decaffeinate the core of the editor (or any other repos/packages in use), a great template for successfully doing so can be seen on [`pulsar-edit/pulsar` PR #13](https://github.com/pulsar-edit/pulsar/pull/13). 
+
+What you must keep in mind is that when Atom was originally created, they mostly used tooling for CoffeeScript 1.x whereas much of the tooling that's available now is CoffeeScript 2.x based. This means that in order to decaffeinate, CoffeeScript made for 1.x, without breaking changes, the tooling for version 1.x **MUST** be used otherwise there may be unknown ramifications. 
+
+But with that said its important to note that not all of Atom's code is CoffeeScript 1.x based. That is to say there is a mix of both used throughout the codebase. In order to accurately determine which is being used within the repo you are working on the best way to do so is to look at that repo's `package.json` `devDependencies` or sometimes it's `dependencies`.
+
+Within those fields look for the `coffee-script` or `coffeescript` package to see if its versions are 1.x or 2.x.
+
+In generally it'll probably be 1.x tooling used.
```
Comment (kaosine) [inline project-birth/CONTRIBUTING-DURING-START.md:87]: This is the only part I'm probably going to mention and maybe right after. This should be "In general" combined with the next part "Now that the version is known,". It just seems odd to me how it's worded here.

**Comment by confused-Techie** [comment]
Alright I've gone ahead and implemented many of the changes suggested by @kaosine @fabianfiorotto and @DeeDeeG 

Think it looks pretty great, but let me know if there's anything that was missed.

**Review by Spiker985 — APPROVED** [review]
Change look good. Verbiage and grammar are good. I can think of nothing more to add.

**Comment by confused-Techie** [comment]
@Spiker985 thanks for the approval, this has been up for some time, with lots of comments from the community. So I'll go ahead and get this merged

**Comment by softcode589** [comment]
Removed @autumnblazey for review because of merge

**Pull request merged**